### PR TITLE
Auto Download Chrome Webdriver

### PR DESCRIPTION
**Pull request by satishkundem**
@Ravieruvuri @mupputur

### CODE REVIEW

File: libUtils/windowsChromeDriver.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import subprocess
+import requests
+import os
+import zipfile
+
+def get_chrome_version():   #    Get the Chrome browser version installed on Windows.
+    try:
+        output = subprocess.check_output(
+            r'reg query "HKEY_CURRENT_USER\Software\Google\Chrome\BLBeacon" /v version',
```
Comment (mupputur) [inline libUtils/windowsChromeDriver.py]: Check the os version

File: libUtils/windowsChromeDriver.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+            encoding='utf-8'
+        )
+        # Extract version from the command output
+        version = output.split()[-1]
+        print(f"Chrome Version: {version}")
+        return version
+
+    except Exception as e:
+        print(f"Error getting Chrome version: {e}")
+        return None
```
Comment (mupputur) [inline libUtils/windowsChromeDriver.py]: raise a customer exception

File: libUtils/windowsChromeDriver.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import subprocess
+import requests
+import os
+import zipfile
+
```
Comment (mupputur) [inline libUtils/windowsChromeDriver.py]: created a class with  BrowserManager   and initialize the browser path and implement respective instance methods inside in it and also your code should extends to other browsers as well

File: libUtils/windowsChromeDriver.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+class BrowserManager:
+    def __init__(self,extract_dir=".//dependencies//chrome-driver"):
```
Comment (mupputur) [inline libUtils/windowsChromeDriver.py]: keep oneline space between each method

Comment (mupputur) [inline libUtils/windowsChromeDriver.py]: change the variable name extract_dir to driver_path

File: libUtils/windowsChromeDriver.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                shell=True,
+                encoding='utf-8'
+            )
+            # Extract version from the command output
+            version = output.split()[-1]
+            print(f"Chrome Version: {version}")
+            return version
+
+        except Exception as e:
+            raise Exception(f"Error getting Chrome version: {e}")
```
Comment (mupputur) [inline libUtils/windowsChromeDriver.py]: reg_path  = HKEY_CURRENT_USER\Software\Google\Chrome\BLBeacon" /v version
Change the exception message to f"Failed to get chrome browser version. Error:  {e}"
Also add print statement before raise statement  f"Failed to read chrome browser version from path:  {reg_path  }"

File: libUtils/windowsChromeDriver.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    def get_chrome_driver_download_url(self, chrome_version):
+        try:
+            if not chrome_version:
```
Comment (mupputur) [inline libUtils/windowsChromeDriver.py]: Removed this condition and invoke this function when there is a chrome version returned from the function get_chrome_version()

File: libUtils/windowsChromeDriver.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                raise Exception("Chrome version is required to fetch the download URL.")
+
+            download_url = f"https://storage.googleapis.com/chrome-for-testing-public/{chrome_version}/win64/chromedriver-win64.zip"
```
Comment (mupputur) [inline libUtils/windowsChromeDriver.py]: moved the base url to inside init method like self.driver_base_url  =f"https://storage.googleapis.com/chrome-for-testing-public/"  and  use it here

File: libUtils/windowsChromeDriver.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            print(f"Chrome Version: {version}")
+            return version
+
+        except Exception as e:
+            raise Exception(f"Error getting Chrome version: {e}")
+
+    def get_chrome_driver_download_url(self, chrome_version):
+        try:
+            if not chrome_version:
+                raise Exception("Chrome version is required to fetch the download URL.")
```
Comment (mupputur) [inline libUtils/windowsChromeDriver.py]: Defined function with name get_driver_uri(): it shoudl return /win64/chromedriver-win64.zip" incase of 64 bit else return corresponding path 32 bit version ,    Also define utility function is_windows64(),  that returns true if return 64 else false

File: libUtils/windowsChromeDriver.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+            download_url = f"https://storage.googleapis.com/chrome-for-testing-public/{chrome_version}/win64/chromedriver-win64.zip"
+            print(f"Download URL: {download_url}")
```
Comment (mupputur) [inline libUtils/windowsChromeDriver.py]: Print statments shoud be more clear :   Chrome Driver Download URL

File: libUtils/windowsChromeDriver.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+            response = requests.head(download_url)
+
+            if response.status_code != 200:
+                raise Exception(
```
Comment (mupputur) [inline libUtils/windowsChromeDriver.py]: Add a print stament before raise and print status code and response  HTTP Status Code: {response.status_code}"

File: libUtils/windowsChromeDriver.py [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    def get_chrome_driver_download_url(self, chrome_version):
+        try:
+            if not chrome_version:
+                raise Exception("Chrome version is required to fetch the download URL.")
+
+            download_url = f"https://storage.googleapis.com/chrome-for-testing-public/{chrome_version}/win64/chromedriver-win64.zip"
+            print(f"Download URL: {download_url}")
+
+            response = requests.head(download_url)
+
```
Comment (mupputur) [inline libUtils/windowsChromeDriver.py]: Explain what this will do ?

File: libUtils/windowsChromeDriver.py [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+                    f"Failed to get ChromeDriver for Chrome {chrome_version}. HTTP Status Code: {response.status_code}")
+
+            return download_url
+        except Exception as e:
+            raise Exception(f"Error getting ChromeDriver download URL: {e}")
+
+    def download_and_extract_chromedriver(self,download_url):
+        """Download and extract the ChromeDriver zip file."""
+        try:
+            response = requests.get(download_url, stream=True)
```
Comment (mupputur) [inline libUtils/windowsChromeDriver.py:43]: separate two differrent function ,   download_zip_file(url) and extract_zip_file()    , get_driver()

File: libUtils/windowsChromeDriver.py [hunk 13]
```diff
@@ -2,45 +2,41 @@
...
+            download_url = f"https://storage.googleapis.com/chrome-for-testing-public/{chrome_version}/{architecture}/chromedriver-{architecture}.zip"
+            print(f"Chrome Driver Download URL: {download_url}")
 
             response = requests.head(download_url)
-
             if response.status_code != 200:
-                raise Exception(
-                    f"Failed to get ChromeDriver for Chrome {chrome_version}. HTTP Status Code: {response.status_code}")
-
+                raise Exception(f"Failed to get ChromeDriver for Chrome {chrome_version}.")
```
Comment (Ravieruvuri) [inline libUtils/windowsChromeDriver.py:34]: Requirement not met: "Add a print stament before raise and print status code and response HTTP Status Code: {response.status_code}"

**Pull request closed without merging**